controllers/machinehealthcheck_controller.go

```diff
@@ -192,3 +265,152 @@ func (r *MachineHealthCheckReconciler) clusterToMachineHealthCheck(o handler.Map
+		return nil
+	}
+
+	var requests []reconcile.Request
+	for k := range mhcList.Items {
+		if r.hasMatchingLabels(&mhcList.Items[k], m) {
+			key := types.NamespacedName{Namespace: mhcList.Items[k].Namespace, Name: mhcList.Items[k].Name}
+			requests = append(requests, reconcile.Request{NamespacedName: key})
+		}
+	}
+	return requests
+}
+
+func (r *MachineHealthCheckReconciler) nodeToMachineHealthCheck(o handler.MapObject) []reconcile.Request {
+	node, ok := o.Object.(*corev1.Node)
+	if !ok {
+		r.Log.Error(errors.New("incorrect type"), "expected a Node", "type", fmt.Sprintf("%T", o))
+		return nil
+	}
+
+	machine, err := r.getMachineFromNode(node.Name)
+	if machine == nil || err != nil {
+		r.Log.Error(err, "Unable to retrieve machine from node", "node", node.GetName())
+		return nil
+	}
+
+	mhcList := &clusterv1.MachineHealthCheckList{}
+	if err := r.Client.List(
+		context.Background(),
+		mhcList,
+		&client.ListOptions{Namespace: machine.Namespace},
+		client.MatchingFields{mhcClusterNameIndex: machine.Spec.ClusterName},
+	); err != nil {
+		r.Log.Error(err, "Unable to list MachineHealthChecks", "node", node.Name, "machine", machine.Name, "namespace", machine.Namespace)
+		return nil
+	}
+
+	var requests []reconcile.Request
+	for k := range mhcList.Items {
+		if r.hasMatchingLabels(&mhcList.Items[k], machine) {
+			key := types.NamespacedName{Namespace: mhcList.Items[k].Namespace, Name: mhcList.Items[k].Name}
+			requests = append(requests, reconcile.Request{NamespacedName: key})
+		}
+	}
+	return requests
+}
+
+func (r *MachineHealthCheckReconciler) getMachineFromNode(nodeName string) (*clusterv1.Machine, error) {
+	machineList := &clusterv1.MachineList{}
+	if err := r.Client.List(
+		context.TODO(),
+		machineList,
+		client.MatchingFields{machineNodeNameIndex: nodeName},
+	); err != nil {
+		return nil, errors.Wrap(err, "failed getting machine list")
+	}
+	if len(machineList.Items) != 1 {
+		return nil, errors.New(fmt.Sprintf("expecting one machine for node %v, got: %v", nodeName, machineList.Items))
+	}
+	return &machineList.Items[0], nil
+}
+
+func (r *MachineHealthCheckReconciler) watchClusterNodes(ctx context.Context, c client.Client, cluster *clusterv1.Cluster) error {
+	key := types.NamespacedName{Namespace: cluster.Name, Name: cluster.Name}
+	if _, ok := r.clusterNodeInformers[key]; ok {
+		// watch was already set up for this cluster
+		return nil
+	}
+
+	config, err := remote.RESTConfig(ctx, c, cluster)
+	if err != nil {
+		return errors.Wrap(err, "error fetching remote cluster config")
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "error constructing remote cluster client")
+	}
+
+	// TODO(JoelSpeed): See if we use the resync period from the manager instead of 0
+	factory := informers.NewSharedInformerFactory(k8sClient, 0)
+	nodeInformer := factory.Core().V1().Nodes().Informer()
+	go nodeInformer.Run(ctx.Done())
```

Comment:
I'd say start the factory vs the individual informer. The net effect is the same, but I think going up to the factory level makes more sense. WDYT?

Reply:
I only started the node informer since we know this is the only informer we will ever use from this factory. If we were to start the factory, does that not then start all informers within the factory, most of which we won't be using?